### PR TITLE
feat(input): add outside-top prop

### LINKED ISSUE #3058: [BUG] - Input - Label with placement "outside" only presented when placeholder === " "
### NextUI Version

2.3.6

### Describe the bug

When using input component i would only like to present the label outside with no place holder. In nextUI there is no option to do so, so if i set labelPlacement="outside" i will have to also set the placeholder=" " to string with spaces. 

image

Code: 
```
 <Input 
        type="email"
        label="Email"
        labelPlacement={"outside"} 
       />

```

image

Code: 
```
   <Input
            type="email"
            label="Email"
            labelPlacement={"outside"}
            placeholder={" "}
          />
```

[Codesandbox](https://codesandbox.io/p/devbox/upbeat-murdock-8s5lld?workspaceId=0bdb6b5c-75b0-41ff-ab06-3dc9acd83d3c&layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clwj11jn700083b6k48y96ws5%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clwj11jn700033b6kxu6sc9b3%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clwj11jn700053b6k15thplwu%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clwj11jn700073b6kba8ak3vr%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clwj11jn700033b6kxu6sc9b3%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clwj11jn700023b6kkx9w4vai%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252FApp.jsx%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522id%2522%253A%2522clwj11jn700033b6kxu6sc9b3%2522%252C%2522activeTabId%2522%253A%2522clwj11jn700023b6kkx9w4vai%2522%257D%252C%2522clwj11jn700073b6kba8ak3vr%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clwj11jn700063b6ktjo0h8bz%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_PORT%2522%252C%2522taskId%2522%253A%2522dev%2522%252C%2522port%2522%253A5173%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522id%2522%253A%2522clwj11jn700073b6kba8ak3vr%2522%252C%2522activeTabId%2522%253A%2522clwj11jn700063b6ktjo0h8bz%2522%257D%252C%2522clwj11jn700053b6k15thplwu%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clwj11jn700043b6k8fiuzskx%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522dev%2522%257D%255D%252C%2522id%2522%253A%2522clwj11jn700053b6k15thplwu%2522%252C%2522activeTabId%2522%253A%2522clwj11jn700043b6k8fiuzskx%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)

### Steps to Reproduce the Bug or Issue

1. Import Input Component
2. Set the follwing props `  type="text"
            label="Name"
            labelPlacement={"outside"}
            placeholder={" "}`

### Expected behavior

should not force to set up placeholder in order to show label outside

### Screenshots or Videos

_No response_

### Operating System Version

MacOS 14.4.1

### Browser

Chrome

### PR DESCRIPTION
**Pull request by abhinav700**
Closes #3058 

## ⛳️ Current behavior (updates)

In old behaviour, the label was presented inside the input component if `placholder === ""` even if `labelPlacement === "outside"` 

## 🚀 New behavior

Adding outside-top prop which dispalys the label outside the input component regardless of whether the placeholder is there or not.

image

## Summary by CodeRabbit

- **New Features**
	- Introduced a new label positioning option, "outside-top," for improved visibility and accessibility in the input component.
	- Enhanced layout options for the Input component with updated styling configurations.

- **Documentation**
	- Updated component stories to showcase the new "outside-top" functionality.
	- Extended label placement options in documentation to include "outside-top."
	- Revised documentation to reflect changes in the `labelPlacement` property.

### CODE REVIEW

File: .changeset/little-crabs-destroy.md
Comment (wingkwong) [inline .changeset/little-crabs-destroy.md:2]: change from `major` to `minor`

File: packages/components/input/src/input.tsx
Comment (wingkwong) [inline packages/components/input/src/input.tsx:52]: what is the reason of adding `bg-green-500` here?

File: packages/components/input/src/input.tsx
Comment (wingkwong) [inline packages/components/input/src/input.tsx:74]: what is the reason of adding `bg-yellow-700` here?

File: packages/components/input/src/input.tsx
Comment (wingkwong) [inline packages/components/input/src/input.tsx:85]: what is the reason of adding `bg-pink-600` here?

File: packages/components/input/src/input.tsx
Comment (wingkwong) [inline packages/components/input/src/input.tsx:97]: what is the reason of adding `bg-blue-900` here?

File: packages/components/input/src/use-input.ts
Comment (wingkwong) [inline packages/components/input/src/use-input.ts:280]: please explain why `!hasStartContent` is removed here

File: packages/components/input/src/use-input.ts
Comment (wingkwong) [inline packages/components/input/src/use-input.ts:271]: add these to `shouldLabelBeOutside` instead

**Review by wingkwong — CHANGES_REQUESTED** [review]

File: packages/components/input/src/input.tsx
Comment (abhinav700) [reply packages/components/input/src/input.tsx:97]: I was testing something and was trying to understand the components.
Looks like I forgot to remove these when I made a commit.
Will change them A.S.A.P.

File: packages/components/input/src/use-input.ts
Comment (abhinav700) [reply packages/components/input/src/use-input.ts:280]: I had uncommented this part while making the commit, but it seeems that code was not saved in VS code.
Will remove the comments.

File: packages/components/input/src/use-input.ts
Comment (abhinav700) [reply packages/components/input/src/use-input.ts:271]: Ok.

File: .changeset/little-crabs-destroy.md
Comment (abhinav700) [reply .changeset/little-crabs-destroy.md:2]: Ok.

**Comment by user72356** [comment]
Can anyone merge this?

**Comment by wingkwong** [comment]
> Can anyone merge this? @jrgarciadev @wingkwong please?

This PR is scheduled for v2.5.0. We will review and merge after v2.4.7.

File: .changeset/little-crabs-destroy.md
Comment (wingkwong) [inline .changeset/little-crabs-destroy.md:6]:
```
introduce `outside-top` prop to input (#3058)
```

File: packages/components/input/src/use-input.ts
Comment (wingkwong) [inline packages/components/input/src/use-input.ts:261]: i got `This comparison appears to be unintentional because the types '"inside" | undefined' and '"outside-top"' have no overlap.ts(2367)`. please double check.

File: packages/components/input/src/use-input.ts
Comment (wingkwong) [inline packages/components/input/src/use-input.ts:269]: i got `This comparison appears to be unintentional because the types '"inside" | undefined' and '"outside-top"' have no overlap.ts(2367)`. please double check.

File: packages/components/input/src/use-input.ts
Comment (wingkwong) [inline packages/components/input/src/use-input.ts:257]: merge the comment below.

```ts
const shouldLabelBeOutside =
    // label is outside only when some placeholder is there
    labelPlacement === "outside" ||
    // label is outside regardless of placeholder
    labelPlacement === "outside-left" ||
    labelPlacement === "outside-top";
```

**Review by wingkwong — CHANGES_REQUESTED** [review]
it looks kinda packed. probably put outsied-left and outside-top in a new row. please also check with docs one also.
image

File: packages/components/input/src/use-input.ts
Comment (abhinav700) [reply packages/components/input/src/use-input.ts:280]: > 2. also i got `This comparison appears to be unintentional because the types '"inside" | undefined' and '"outside-top"' have no overlap.ts(2367)`. please double check.

I don'y see this message in IDE or chrome dev tools. can you tell me where you got this?

Comment (abhinav700) [reply packages/components/input/src/use-input.ts:280]: > it looks kinda packed. probably put outsied-left and outside-top in a new row. please also check with docs one also. 

is this okay?
image

File: packages/core/theme/src/components/input.ts
Comment (abhinav700) [reply packages/core/theme/src/components/input.ts:177]: > in `packages/components/input/package.json`, currently the peerDependencies of `@nextui-org/theme` is `>=2.1.0`. Need to bump that version.

What should I change it to?

File: packages/core/theme/src/components/input.ts
Comment (abhinav700) [reply packages/core/theme/src/components/input.ts:177]: should I add this change leave it as it is?

File: packages/components/input/src/use-input.ts
Comment (wingkwong) [reply packages/components/input/src/use-input.ts]: looks good

File: packages/core/theme/src/components/input.ts
Comment (wingkwong) [reply packages/core/theme/src/components/input.ts:177]: you may update to the next target version of `@nextui-org/theme`

Comment (abhinav700) [reply packages/core/theme/src/components/input.ts:177]: 2.2.0?

File: packages/core/theme/src/components/input.ts
Comment (wingkwong) [reply packages/core/theme/src/components/input.ts:177]: image

may double check why the label in outside one is larger than others. ref: https://nextui-storybook-v2-mwmsxwqw8-nextui-org.vercel.app/?path=%2Fstory%2Fcomponents-input--label-placement&args=size%3Alg

File: packages/core/theme/src/components/input.ts
Comment (wingkwong) [reply packages/core/theme/src/components/input.ts:177]: currently the version of `@nextui-org/theme` is `2.2.11`. so next minor version would be `2.3.0`.

File: packages/core/theme/src/components/input.ts
Comment (abhinav700) [reply packages/core/theme/src/components/input.ts:177]: Okay, will look into that

File: packages/core/theme/src/components/input.ts
Comment (abhinav700) [reply packages/core/theme/src/components/input.ts:177]: okay

File: packages/core/theme/src/components/input.ts
Comment (abhinav700) [reply packages/core/theme/src/components/input.ts:177]: I have fixed that in my latest commit.
it was because of the `text-medium` class present for `lg` and `outside` compound variant.

**Review by wingkwong — COMMENTED** [review]
seems this is only useful for "without placeholder". for "with placeholder, basically this prop is same as outside. I'm wondering if this would confuse users or not. @ryo-manba what do you think?

image

**Comment by wingkwong** [comment]
marking this PR as `on hold` until the docs structure revamp pr merged to beta branch.

**Comment by abhinav700** [comment]
> seems this is only useful for "without placeholder". for "with placeholder, basically this prop is same as outside. I'm wondering if this would confuse users or not. @ryo-manba what do you think?

to resolve this, I think we can replace `outside` with  something `outside-conditional` or `outside-partial`

**Comment by ryo-manba** [comment]
@wingkwong 
I certainly see your point, but I think this is fine since it matches the `outside-left` behavior.
It might also be helpful to add a note in the docs that both `outside-left` and `outside-top` always display labels outside, even without a placeholder.

**Comment by wingkwong** [comment]
> It might also be helpful to add a note in the docs that both outside-left and outside-top always display labels outside, even without a placeholder.

@abhinav700 please pull the latest code and add this note. Then we'll have a final review. Thanks.

**Comment by abhinav700** [comment]
@wingkwong Done. I have pulled the latest code and added the notes for `outside-top` and `outside`
image

**Comment by abhinav700** [comment]
@wingkwong I have pulled the code and updated the docs

**Pull request closed without merging**
